### ISSUE TITLE
An Item should know its inspection questions

### ISSUE THREAD

**Issue by kgf** [ReadyResponder/ReadyResponder#487]
Currently an item can have an inspection, but it's kind of lame. All that's asked is the date and the current condition. What would be good would be to have a dynamic list of questions.
I would like the list of questions to be defined for each item_category, and also each item_type. In the future, an item may have more than one item_type in a tree structure. (An item_type may have other item_types as children.) The questions should be collected up and presented to the user on the inspection form.
For this iteration, all questions can be yes/no.

If this is too large for one PR, it could be broken into several steps. The first would be to have `@item.inspection_questions` return a collection of questions.

**Comment by kgf** [comment]
@tegon, would this be of interest ?

**Comment by tegon** [comment]
@kgf This one seems too big, I don't think I'll have time to finish it. But thanks anyway.

**Comment by kgf** [comment]
@tegon thanks for your interest. Let me know if you'd like me to come up with others. It's an active project.

**Comment by ChanChar** [comment]
@kgf I've added the `inspection_questions` association but going to need some clarification the rest of the task.

**Comment by kgf** [comment]
I think that may be the wrong plan. The inspection questions will need to be on ItemCategory and ItemType. An Item will look up it's hierarchy, and gather all the questions to put on the Inspection. An Inspection instance will have something like `inspection_answers`, which should include the text of the original question, and the response. The first iteration can have simple pass/fail yes/no questions, like "Are the tires properly inflated ?" and "Do all the exterior lights work ?"
The answer should have more than a pointer to the question, in case someone edits the text of the question to have a different meaning.
Hope this helps.
Thanks
Kevin

**Comment by kgf** [comment]
@ChanChar Any updates ? Anything I can help with ?

**Comment by ChanChar** [comment]
@kgf Where are these questions coming from? 

> The answer should have more than a pointer to the question, in case someone edits the text of the question to have a different meaning.

Shouldn't we simply remove the associated answer and have the user reset the answer if this is a concern?

All in all, I may not have time to do this and will leave it open for someone else.

**Comment by kgf** [comment]
OK, no problem. I'll just reset the labels. Thanks for your help in the past !

**Comment by nsantiago2719** [comment]
Hi @kgf , I think I can take this. But for just clarifications, what you want is, the user should be able to add a dynamic question on each question right?

**Comment by kgf** [comment]
@nsantiago2719 Thanks for your interest. 
I don't follow quite what you mean. Be sure to read my comment from ten days ago. 

The admin user would add questions to `item_categories` and `item_types`. When any user begins an inspection, the system will find the Items type, and collect all the questions defined for that `item_type`. It will then find the item_type's item_category, and collect all the questions defined for that item_category.
In production, `Vehicles` is a popular item_category. All vehicles need to be inspected for proper tire pressure. So that would be a question defined on the Item_Category. Then, under vehicles, you would have an item_type of `Pick-up Truck`. You might then define a question here like "Is the trailer hitch in good working order". That's not a question appropriate for cars, they don't have trailer hitches. So now, if someone is inspecting vehicle A-10, which has an `item_type` of Pickup, and an `item-category` of vehicle, the inspection questions will be `Tire pressure OK ?` and `Hitch in good working order.` The answers will be collected in `inspection_answers`, which will `belong_to` an inspection, and will capture both the answer and the question.
I hope all that helps.

**Comment by kgf** [comment]
@nsantiago2719 Did you have any success with this ? Do you have further questions ?

**Comment by nsantiago2719** [comment]
Hi @kgf , been in a busy week lately. I think I get what you're saying already, I re-read your comments above and I run the project here in my local machine.

**Comment by kgf** [comment]
Thanks a lot for checking in, and letting me know you're still interested. If you have any other questions, please ask !

Thanks for your help,
Kevn

**Comment by kgf** [comment]
@nsantiago2719 has your schedule eased up any? Do you think you'll still be able to help us with this issue?

**Comment by nsantiago2719** [comment]
Hi @kgf,

My schedule is still on tight spot yet. I started working on it but you can open it to others because they might have a better time than mine. Thanks

**Comment by kgf** [comment]
Ok, thanks for the update

**Comment by saltmade** [comment]
Hi @kgf. Not sure if I'm the right skill level, but perhaps if we talk through it I can take it on. So to summarize what's written above.
```
- Category
  - Item Type 1
    - Item Type A
      - Item
```
To run an inspection on `Item` we would pull all the questions from `Item type A`, `Item type 1`, and `Category` into a single form. I would assume duplicate questions would be removed. A `question`s table would allow us to stay dry and de-dupe. You bring up a good point though, how to deal with questions that change so that the answers no longer make sense? 

I can see a couple of approaches: 
1. We could create new instances of a `question` when a `question` is edited. This might create some UI complexity but feels simpler to implement.
2. We could store the `question` with the `answer`. Do you see any future need for tallying up answers to questions or a need to reference a database of questions? If not, this could work.

p.s. I assume that tree-like approach is the only way an `Item` can have multiple `Types` and not
```
- Category
  - Item Type 1
    - Item A
  - Item Type 2
    - Item A
```
And that `Items` and `Item Type`s can't be siblings?
```
- Category
  - Item Type
    - Item
    - Item Type
      - Item
```

**Comment by kgf** [comment]
I feel a variant of #1 is the correct answer. You can only edit a question to change it's status to inactive, thus all the existing answers remain valid. Perhaps a bit of help directly on the edit form would help.

I would like to be able to have Items and Item_types as siblings, but it's not a requirement.

Good questions !

**Comment by kgf** [comment]
Regarding the structure of items, you may want to look at #539.

**Comment by saltmade** [comment]
@kgf That's a good call. I think #539 is going to be a blocker for this one, given we might use a different gem or approach to move up the tree. I'll pop over there.

Edit: Didn't see that #539 was open again, so perhaps this will set the precedent for how the tree is navigated. I'll look into the Ancestry gem you mentioned there.

**Comment by saltmade** [comment]
@kgf Instead of a tree, would it perhaps make more sense to take a tagging approach to Item_types? I haven't seen the real world examples of what the tree would look like, but from a UX perspective it might mean less burden on the user. If a tree is what fits their mental model, that's fine, but I thought I'd ask before committing to a solution.

**Comment by kgf** [comment]
Interesting ! So you thinking we would tag questions and items, and collect the matches. So we might have `Engine` tag, with `Check Oil` and then tag cars and generators with `Engine` ?

**Comment by jefheaton** [comment]
That's not what I was thinking originally, but that could make sense! If there's enough crossover between categories. I was thinking of treating Item_type as a tag, but now that I'm plotting it out I'm not sure that makes sense. 

Ok, taking a crack. Thanks for the timely response!

**Comment by VanAuken** [comment]
Howdy @jefheaton!
@kgf sent me over to look at your suggestion.  My knee jerk response was negative, but the longer I think about applying it, the better I like it.  
Your tagging approach will be far more flexible.  This is useful because the inventory we're supporting has a lot of variety, both in types of items and variations within a type.  [An example is that we have four or five different models of generators.]
I would like to see an item inherit **tags** from its **category** and **item type** which can then be edited as needed if inappropriate.  We can also create bunches of questions which are re-used across several item types.
It will be important to capture the wording of the questions with the answers.  Later questions may be reworded and we will we want old inspections' answers to make sense, which they may not when compared to newer questions.  
I would like the possible first draft answers to questions to have three options.  They basically boil down to 'Operational', 'Repaired', 'Repair Deferred'.  They can be worded differently, but I found that those responses reflect reality better than yes / no.

**Comment by saltmade** [comment]
@VanAuken Oh excellent! Okay, so the questions could be stored as strings in an array with the answers, but by doing that we make it much harder to look across all inspections for the use of that question (we could probably do that with PG's fulltext search). I _assume_ that doesn't really matter, we probably wouldn't join these up into some kind of report, but I haven't tackled this kind of problem before so I don't have the benefit of knowing for sure.

Alternatively we could keep all questions ever written and whenever a question is edited successfully we just treat it as a new question and replace references to the old one. Old inspections would point to the original question to make sure they still make sense. This means that even simple spelling errors would be maintained, but answers still match their questions and we can run queries against these questions more easily.

Thoughts? 

👍 on the new answers to questions

**Comment by kgf** [comment]
I was thinking the Questions would be like templates. Creating an inspection would create a lot of inspection_questions instances, which would include the answers. I don't expect to ever care about all the times we've asked if the oil is filled. But I would care about the last time someone checked the oil.

@VanAuken What about skipped questions ? 'Operational', 'Repaired', 'Repair Deferred', "Skipped" ?

**Comment by saltmade** [comment]
So each inspection_question record would include both the question and the answer. Makes sense.

**Comment by VanAuken** [comment]
While I'm not expecting the need for a document trail for these inspections, I could imagine an application of the software where the inspection record was necessary to show due diligence and scheduled maintenance. Ensuring that the legacy documents created by the inspection are coherent is never a bad idea.

**Comment by kgf** [comment]
@jefheaton Will you still be able to help us with this ?
Thanks,
Kevin

**Comment by saltmade** [comment]
@kgf Yup! Added a WIP PR here https://github.com/ReadyResponder/ReadyResponder/pull/682. Sorry for the delay, my internship has been busy.

**Comment by kgf** [comment]
@jefheaton Do you still have any interest in this ? Hacktoberfest is upon us...
